slack-lint-checks/src/main/java/slack/lint/JsonInflaterMoshiCompatibilityDetector.kt

```diff
@@ -0,0 +1,160 @@
+// Copyright (C) 2025 Slack Technologies, LLC
+// SPDX-License-Identifier: Apache-2.0
+package slack.lint
+
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import com.android.tools.lint.detector.api.TextFormat
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiModifier
+import com.intellij.psi.PsiPrimitiveType
+import com.intellij.psi.PsiType
+import org.jetbrains.uast.UCallExpression
+import slack.lint.util.sourceImplementation
+
+/**
+ * A detector that checks if a type passed into the JsonInflater.inflate/deflate methods follows
+ * Moshi's requirements for serialization/deserialization.
+ */
+class JsonInflaterMoshiCompatibilityDetector : Detector(), SourceCodeScanner {
+
+  override fun getApplicableUastTypes() = listOf(UCallExpression::class.java)
+
+  override fun createUastHandler(context: JavaContext): UElementHandler {
+    return object : UElementHandler() {
+      override fun visitCallExpression(node: UCallExpression) {
+        if (isJsonInflaterInflateOrDeflate(node)) {
+          when (node.methodName) {
+            "inflate" -> validateInflateReturnType(context, node)
+            "deflate" -> validateDeflateArguments(context, node)
+          }
+        }
+      }
+    }
+  }
+
+  private fun isJsonInflaterInflateOrDeflate(node: UCallExpression): Boolean {
+    // Get the method being called
+    val method = node.resolve() ?: return false
+
+    // Check if it's from JsonInflater class
+    val containingClass = method.containingClass?.qualifiedName ?: return false
+    if (containingClass != FQCN_JSON_INFLATER) return false
+
+    // Check if it's an inflate or deflate method
+    return method.name == "inflate" || method.name == "deflate"
+  }
+
+  private fun validateInflateReturnType(context: JavaContext, node: UCallExpression) {
+    // Get the return type of the inflate call
+    val returnType = node.getExpressionType() ?: return
+
+    // Skip checking primitive types and String
+    if (isPrimitiveOrString(returnType)) return
+
+    // Get the class for the return type
+    val returnClass = context.evaluator.getTypeClass(returnType) ?: return
+
+    // Validate if the class is Moshi-compatible
+    validateClassForMoshiCompatibility(context, node, returnClass)
+  }
+
+  private fun isPrimitiveOrString(type: PsiType): Boolean {
```

Comment:
maybe rename to `isJsonPrimitive`?

Reply:
[4a99e54](https://github.com/slackhq/slack-lints/pull/387/commits/4a99e5460ab0df1183fc6d02acd3fe5d70d2dda5)